slsReceiverSoftware/src/MasterAttributes.h

```diff
@@ -0,0 +1,567 @@
+#pragma once
+
+#include "ToString.h"
+#include "logger.h"
+#include "sls_detector_defs.h"
+
+#ifdef HDF5C
+#include "H5Cpp.h"
+#ifndef H5_NO_NAMESPACE
+using namespace H5;
+#endif
+#endif
+
+#include <chrono>
+using ns = std::chrono::nanoseconds;
+
+// versions
+#define HDF5_WRITER_VERSION   (6.1) // 1 decimal places
+#define BINARY_WRITER_VERSION (6.1) // 1 decimal places
+
+class MasterAttributes {
+  public:
+    slsDetectorDefs::detectorType detType{slsDetectorDefs::GENERIC};
+    slsDetectorDefs::timingMode timingMode{slsDetectorDefs::AUTO_TIMING};
+    uint32_t imageSize{0};
+    slsDetectorDefs::xy nPixels{};
+    uint32_t maxFramesPerFile{0};
+    uint64_t totalFrames{0};
+    ns exptime{0};
+    ns period{0};
+    uint32_t dynamicRange{0};
+    uint32_t tenGiga{0};
+    int threshold{0};
+    ns subExptime{0};
+    ns subPeriod{0};
+    uint32_t quad{0};
+    std::vector<int64_t> ratecorr;
+    uint32_t adcmask{0};
+    uint32_t analog{0};
+    uint32_t digital{0};
+    uint32_t dbitoffset{0};
+    uint64_t dbitlist{0};
+    slsDetectorDefs::ROI roi{};
+    ns exptime1{0};
+    ns exptime2{0};
+    ns exptime3{0};
+    ns gateDelay1{0};
+    ns gateDelay2{0};
+    ns gateDelay3{0};
+    uint32_t gates;
+
+    MasterAttributes(){};
+    virtual ~MasterAttributes(){};
+
+    virtual void WriteMasterBinaryAttributes(FILE *fd) {
```

Comment:
By making this a pure virtual function we can at compile time protect against this since you can't create a base class object. Then we don't need the error message.

Reply:
they are not implemented in the hdf5 part and so cant make it pure virtual